File: gitspaces/space.go
package gitspaces

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"slices"
	"strings"

	"github.com/davfive/gitspaces/v2/console"
	"github.com/davfive/gitspaces/v2/helper"

	cp "github.com/otiai10/copy"
)

type SpaceStat struct {
	space      *SpaceStruct
	wasStated  bool
	IsEmpty    bool
	numRepos   int
}

func NewSpaceStat(space *SpaceStruct) *SpaceStat {
	return &SpaceStat{
		space: space,
	}
}

func (stat *SpaceStat) stat() *SpaceStat {
	if stat.IsEmpty, err helper.DirectoryIsEmpty(space.Path) {
		
}


// Gitspace is a struct that represents a git repository
type SpaceStruct struct {
	Name       string
	Path       string
	Stat       *SpaceStat
	IsEmpty    bool
	IsMonoRepo bool
	project    *ProjectStruct
	codeWsFile string
}

func CreateSpace(project *ProjectStruct) *SpaceStruct {
	newSpace := NewSpace(project, project.getEmptySleeperPath())
	newSpace.init()
	return newSpace
}

func (space *SpaceStruct) spaceState() (int, error) {
	// Assumes space directory exists
	if isEmpty, err := helper.DirectoryIsEmpty(space.Path); err != nil {
		return -1, err
	}

	if isEmpty {
		return 0, nil
	}

	if helper.PathExists(filepath.Join(space.Path, ".git")) {
		return 1
	}

	

	if _, err := os.Stat(filepath.Join(space.Path, ".git")); !os.IsNotExist(err) {

	os.ReadDir(space.Path)

	return helper.CountDirs(space.Path)
}


func (space *SpaceStruct) AddRepoFromUrls(project *ProjectStruct, url string, path string) error {
	dir := strings.TrimSuffix(filepath.Base(url), ".git")

	if space.numRepos() > 0 {
		return errors.New("space already has a repo")
	}

	if _, err := os.Stat(filepath.Join(space.Path, dir)); err == nil {
		return errors.New("repo dir already exists in space")
	}

	space = NewSpace(project, path)
	urls = slices.Compact(urls)
	urlMap := make(map[string]string)
	for _, url := range urls {
		dir := strings.TrimSuffix(filepath.Base(url), ".git")
		if _, ok := urlMap[dir]; ok {
			urlMap[dir] = url
		} else {
			// TODO: We could ask the user if they want to use a different name for the duplicate
			return nil, errors.New("duplicate repo name in urls list: " + strings.Join(urls, ","))
		}
	}

	getRepoDir := func(dir string) string {
		if len(urls) == 1 {
			// single-repo spaces are cloned directly into the space directory
			// Makes the path shorter for simple projects
			return path
		}
		return filepath.Join(path, dir)
	}

	// go-git doesn't have a robust ssh-cloning implementation (gits tripped up easily by ssh config, )
	for dir, url := range urlMap {
		cmd := exec.Command("git", "clone", url, getRepoDir(dir))
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		if err = cmd.Run(); err != nil {
			return nil, err
		}
	}

	if err = space.createCodeWorkspaceFile(); err != nil {
		return nil, err
	}

	return space, nil
}

func GetSpaceFromPath(path string) (*SpaceStruct, error) {
	project, err := GetProjectFromPath(path) // Path is checked here
	if err != nil {
		return nil, err
	}

	for _, spacePath := range project.getWorkerSpacePaths() {
		if strings.HasPrefix(path, spacePath) {
			return NewSpace(project, spacePath), nil
		}
	}

	return nil, errors.New("space not found from " + path)
}

func GetSpace() (*SpaceStruct, error) {
	return GetSpaceFromPath(helper.Getwd())
}

// NewSpace creates a new Space struct
func NewSpace(project *ProjectStruct, path string) *SpaceStruct {
	space := &SpaceStruct{
		Path:    path,
		project: project,
	}
	space.Stat = NewSpaceStat(space)
	space.updateName()
	return space
}

func (space *SpaceStruct) Duplicate() (newSpace *SpaceStruct, err error) {
	copyToPath := space.project.getEmptySleeperPath()

	if err = cp.Copy(space.Path, copyToPath); err != nil {
		return nil, err
	}

	return NewSpace(space.project, copyToPath), nil
}

func (space *SpaceStruct) GetProject() *ProjectStruct {
	return space.project
}

func (space *SpaceStruct) OpenVSCode() error {
	codeExecPath, err := exec.LookPath("code")
	if err != nil {
		return err
	}
	return exec.Command(codeExecPath, space.codeWsFile).Start()
}

// Rename renames the space.
// It takes a variadic parameter of type string, which represents the optional new name for the space.
// If successful, it returns nil. Otherwise, it returns an error.
func (space *SpaceStruct) Rename(arguments ...string) error {
	return space.move("Rename", arguments...)
}

func (space *SpaceStruct) Sleep() (err error) {
	space.deleteCodeWorkspaceFile()

	newPath := space.project.getEmptySleeperPath()
	space.Name = filepath.Base(space.Path)
	if err = os.Rename(space.Path, newPath); err != nil {
		return err
	}
	space.deleteCodeWorkspaceFile()

	return nil
}

func (space *SpaceStruct) Wakeup() (err error) {
	return space.move("Wakeup")
}

func (space *SpaceStruct) createCodeWorkspaceFile() (err error) {
	var file *os.File

	// Ensure it has the correct name
	space.codeWsFile = filepath.Join(
		space.project.codeWsDir,
		fmt.Sprintf("%s~%s.code-workspace", space.project.Name, space.Name),
	)

	if file, err = os.Create(space.codeWsFile); err != nil {
		return err
	}
	defer file.Close()

	_, err = file.WriteString(fmt.Sprintf(`{
		"folders": [
			{
				"path": "%s"
			}
		],
		"settings": {}
	}`, space.Path))

	return err
}

func (space *SpaceStruct) deleteCodeWorkspaceFile() (err error) {
	if helper.PathExists(space.codeWsFile) {
		if err = os.Remove(space.codeWsFile); err != nil {
			fmt.Fprintln(os.Stderr, "Failed to remove "+space.codeWsFile)
			return err
		}
	}
	return nil
}

func (space *SpaceStruct) init() error {
	// Minimal setup for space, just create the space directory
	return os.MkdirAll(space.Path, os.ModePerm)
}

func (space *SpaceStruct) move(moveVerb string, arguments ...string) error {
	var newName string // pining for default args and/or a ternary operator :/
	if len(arguments) > 0 {
		newName = arguments[0]
	}
	name, err := console.GetUserInputWithValidation(
		fmt.Sprintf("%s space as", moveVerb),
		newName,
		console.MakeDirnameAvailableValidator(space.project.Path),
	)
	if err != nil {
		return err
	}

	newPath := filepath.Join(space.project.Path, name)
	space.deleteCodeWorkspaceFile()
	if err = os.Rename(space.Path, newPath); err != nil {
		return err
	}
	space.Path = newPath
	space.Name = name
	if err = space.createCodeWorkspaceFile(); err != nil {
		fmt.Fprintln(os.Stderr, "Failed to create code workspace file")
	}

	return nil
}

func (space *SpaceStruct) updateName() {
	// It would just be filepath.Base(space.path) but we have to account for sleeper spaces (in .zzz dir)
	// The space.Name is used in the code-workspace file name so it can't have any separator characters
	space.Name = strings.TrimPrefix(space.Path, space.project.Path+string(filepath.Separator))
	space.Name = strings.ReplaceAll(space.Name, string(filepath.Separator), "~")
}

File: src/gitspaces/modules/space.py
"""Space management for GitSpaces."""

from pathlib import Path
from typing import Optional
from git import Repo
from gitspaces.modules.errors import SpaceError
from gitspaces.modules.path import ensure_dir
from gitspaces.modules import runshell


class Space:
    """Represents a single workspace (clone) within a GitSpaces project."""

    def __init__(self, project, path: str):
        """Initialize a Space.

        Args:
            project: The parent Project instance.
            path: The path to the space directory.
        """
        self.project = project
        self.path = Path(path)
        self.name = self.path.name
        self._repo: Optional[Repo] = None

    @property
    def repo(self) -> Optional[Repo]:
        """Get the git repository for this space.

        Returns:
            The GitPython Repo instance or None.
        """
        if self._repo is None:
            self._repo = runshell.git.get_repo(str(self.path))
        return self._repo

    @classmethod
    def create_space_from_url(cls, project, url: str, path: str) -> "Space":
        """Create a new space by cloning from a URL.

        Args:
            project: The parent Project instance.
            url: The git repository URL.
            path: The path where the space will be created.

        Returns:
            The created Space instance.
        """
        if Path(path).exists():
            raise SpaceError(f"Space directory already exists: {path}")

        runshell.git.clone(url, path)
        space = cls(project, path)
        return space

    def duplicate(self) -> "Space":
        """Duplicate this space to a new sleeper space.

        Returns:
            The new Space instance.
        """
        new_path = self.project._get_empty_sleeper_path()

        try:
            # Copy the entire directory
            runshell.fs.copy_tree(str(self.path), new_path, symlinks=True)
        except Exception as e:
            raise SpaceError(f"Failed to duplicate space: {e}")

        return Space(self.project, new_path)

    def wake(self, new_name: Optional[str] = None) -> "Space":
        """Wake up a sleeping space and optionally rename it.

        Args:
            new_name: Optional new name for the space.

        Returns:
            The woken Space instance (may be a new instance if renamed).
        """
        # Check if this is a sleeping space
        if not str(self.path).startswith(str(self.project.zzz_dir)):
            raise SpaceError("Space is not sleeping")

        # Determine the new path
        if new_name:
            new_path = self.project.path / new_name
        else:
            # Use the default branch name or 'main'
            repo = self.repo
            if repo:
                branch_name = runshell.git.get_active_branch(repo)
            else:
                branch_name = "main"
            new_path = self.project.path / branch_name

        if new_path.exists():
            raise SpaceError(f"Target directory already exists: {new_path}")

        # Move the space
        runshell.fs.move(str(self.path), str(new_path))

        return Space(self.project, str(new_path))

    def sleep(self) -> "Space":
        """Put this space to sleep (move to .zzz directory).

        Returns:
            The sleeping Space instance.
        """
        # Check if already sleeping
        if str(self.path).startswith(str(self.project.zzz_dir)):
            raise SpaceError("Space is already sleeping")

        new_path = self.project._get_empty_sleeper_path()

        # Move the space
        runshell.fs.move(str(self.path), new_path)

        return Space(self.project, new_path)

    def rename(self, new_name: str) -> "Space":
        """Rename this space.

        Args:
            new_name: The new name for the space.

        Returns:
            The renamed Space instance.
        """
        # Check if sleeping
        if str(self.path).startswith(str(self.project.zzz_dir)):
            new_path = self.project.zzz_dir / new_name
        else:
            new_path = self.project.path / new_name

        if new_path.exists():
            raise SpaceError(f"Target directory already exists: {new_path}")

        # Rename the space
        runshell.fs.move(str(self.path), str(new_path))

        return Space(self.project, str(new_path))

    def get_current_branch(self) -> str:
        """Get the current branch name.

        Returns:
            The current branch name or "detached".
        """
        repo = self.repo
        if repo:
            return runshell.git.get_active_branch(repo)
        return "detached"

    def is_sleeping(self) -> bool:
        """Check if this space is sleeping.

        Returns:
            True if the space is in the .zzz directory.
        """
        return str(self.path).startswith(str(self.project.zzz_dir))
